crates/ty_python_semantic/src/types/class.rs

```diff
@@ -2516,17 +2516,20 @@ impl<'db> ClassLiteral<'db> {
                 //     self.name: <annotation>
                 //     self.name: <annotation> = …
 
-                if use_def_map(db, method_scope)
-                    .is_declaration_reachable(db, &attribute_declaration)
-                    .is_always_false()
-                {
+                let reachability = use_def_map(db, method_scope)
+                    .is_declaration_reachable(db, &attribute_declaration);
+
+                if reachability.is_always_false() {
                     continue;
                 }
 
                 let annotation = declaration_type(db, declaration);
-                let annotation =
+                let mut annotation =
                     Place::bound(annotation.inner).with_qualifiers(annotation.qualifiers);
 
+                if reachability.is_ambiguous() {
+                    annotation.qualifiers |= TypeQualifiers::NOT_BOUND;
+                }
```

Comment:
The lines here only handle *declared* instance attributes (like `self.x: Literal[True] = True`) in the `Toggle` example. We should do the same for non-declared instance attributes (e.g. `self.x = True`). This modified `Toggle` example shows that this is currently still problematic on this branch:
```py
from typing import Literal

def check(x) -> Literal[False]:
    return False

class Toggle:
    def __init__(self: "Toggle"):
        if check(self.x):
            self.x = True
```

I had to use a little trick here: `if not self.x` wouldn't work anymore when `self.x` is not declared. For undeclared symbols, we infer `Unknown | True`, and the `Unknown` will not lead to a definitive Boolean answer. So I added this `check(…)` function that always returns `Literal[False]`.